plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/modelsourcewatcher/FileWatcher.java

```diff
@@ -0,0 +1,195 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+package org.eclipse.glsp.server.features.modelsourcewatcher;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.ClosedWatchServiceException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.glsp.server.actions.ActionDispatcher;
+import org.eclipse.glsp.server.model.GModelState;
+import org.eclipse.glsp.server.protocol.ClientSessionListener;
+import org.eclipse.glsp.server.protocol.ClientSessionManager;
+import org.eclipse.glsp.server.protocol.GLSPClient;
+import org.eclipse.glsp.server.utils.ClientOptions;
+
+import com.google.inject.Inject;
+
+public class FileWatcher implements ClientSessionListener, ModelSourceWatcher {
+
+   private static final String FILE = "file://";
+
+   @Inject
+   private ActionDispatcher actionDispatcher;
+
+   private final Map<String, List<FileWatchWorker>> workers = new HashMap<>();
+
+   @Inject
+   public FileWatcher(final ClientSessionManager sessionManager) {
+      sessionManager.addListener(this);
+   }
+
+   public FileWatcher(final ClientSessionManager sessionManager, final ActionDispatcher actionDispatcher) {
+      this(sessionManager);
+      this.actionDispatcher = actionDispatcher;
+   }
+
+   @Override
+   public void sessionClosed(final String clientId, final GLSPClient client) {
+      disposeAllWorkers(clientId);
+   }
+
+   @Override
+   public void startWatching(final GModelState modelState) {
+      createWorkers(modelState).forEach(FileWatchWorker::start);
+   }
+
+   @Override
+   public void stopWatching(final GModelState modelState) {
+      disposeAllWorkers(modelState.getClientId());
+   }
+
+   @Override
+   public void pauseWatching(final GModelState modelState) {
+      getAllWorkers(modelState.getClientId()).forEach(FileWatchWorker::pauseNotifications);
+   }
+
+   @Override
+   public void continueWatching(final GModelState modelState) {
+      getAllWorkers(modelState.getClientId()).forEach(FileWatchWorker::continueNotifications);
+   }
+
+   protected List<Path> getPaths(final GModelState modelState) {
+      final Optional<String> uriString = ClientOptions.getValue(modelState.getClientOptions(),
+         ClientOptions.SOURCE_URI);
+      URI sourceUri = URI.createFileURI(uriString.map(uri -> uri.replace(FILE, "")).orElseThrow());
```

Comment:
I am not a big fan of this URI manipulation hidden in this method. There is something similar going on in the JsonFileModelFactory (convertToFile). Do you think it would make sense to extract that into a utility class somewhere or even put it into the ClientOptions as a special method since they already define the constants for that anyway. Same goes for the FILE constant.

Reply:
Yes, I agree! I added a static method to ClientOptions and harmonized those two methods.